Implement Tensordot

amjames opened the pull request:
This PR adds an implementation of tensordot. 

`numpy.tensordot` accepts two ways of specifying the `axes` argument. 
One where axes is an integer and specifies that the last (first) `axes` axes in `a` (`b`) will be those summed over. The other is where axes is a pair of tuples of the same length and the axes from the first tuple are those summed over for `a` and the axes from the second tuple are those summed over for `b`. 

## Small difference between numpy's syntax and xtensor's syntax.

I have implemented this as two overloads one, taking the single `naxes` argument which default to 2 (same as numpy), and the second overload taking two vectors, one for a and one for b. The resulting syntax is similar in appearance to that of numpy with one small exception. 

### Axes is a single integer:
No difference here. 
numpy: 
```python
np.tensordot(a, b) # default behavior axes = 2
np.tensordot(a, b, axes=3) # specify axes != 2
```
xtensor:
```c++
xt::linalg::tensordot(a, b) // calls first overload, default behavior naxes = 2
xt::linalg::tensordot(a, b, 3) // calls first  overload, specify naxes != 2
```

### Axes is specified using vectors/tuples (numpy/xtensor):
Here is the difference. In numpy axes is still one argument.
```python
np.tensordot(a, b, axes=((1,3),(2,3))) # axes is a pair of tuples
```
In my implementation for xtensor, axes is split into two arguments
```c++
xt::linalg::tensordot(a, b, {1,3}, {2,3})
```

Happy to add additional tests, or respond to any feedback about performance problems with my implementation (I was learning how the lazy evaluation of expressions works while I was writing this, so I am sure that I was over cautious with copies/forcing evaluation and there is probably room for performance improvements).

wolfv commented:
Wow, this is great! Will give it a thorough review tomorrow!

tdegeus commented:
This seems really great to have! Do you by any chance have ideas for an `einsum` equivalent?

amjames commented:
@tdegeus In fact, I had originally set out to implement `einsum`, but I was not really sure what the best way to deal with the interface would be. After scratching my head about that for a while I just went ahead with `tensordot` which covers a lot of the functionality `einsum` would be used for, namely tensor contractions. With the addition of `tensordot` `xtensor`/`xtensor-blas` would now provide separate implementations of all the individual operations that `einsum` can be used for. 

I have thought about it though, and after summarizing those thoughts I think that such a discussion may be more appropriate in a separate issue. Shall I open one?

wolfv commented on include/xtensor-blas/xlinalg.hpp:
```diff
@@ -1514,6 +1514,172 @@ namespace linalg
+        if (naxes == 0)
+        {
+            // special case tensor outer product product
+            std::vector<std::size_t> result_shape(a.dimension() + b.dimension());
```
it could be a good idea to use the `xt::dynamic_shape` typedef. The reason is that the `dynamic_shape`, by default, is a `svector`, which has a small buffer optimization and doesn't require allocation if the vector size stays < 4.

amjames replied:
Excellent, didn't know about `dyamic_shape` I fixed this.

wolfv commented on include/xtensor-blas/xlinalg.hpp:
```diff
@@ -1514,6 +1514,172 @@ namespace linalg
+                // check for axes size match
+                if (a_val != b_val)
+                {
+                    throw std::runtime_error("Shape mismatch for sum\n");
```
I'm not sure about the `\n` at the end of the message. Are we doing this elsewhere?

amjames replied:
Looking at other places in `xlinalg.hpp` it appears no. I will remove it.

wolfv commented on include/xtensor-blas/xlinalg.hpp:
```diff
@@ -1514,6 +1514,172 @@ namespace linalg
+                ++bs_it;
+            }
+            std::vector<std::size_t> result_shape;
+            auto r_shape_inserter = std::back_inserter(result_shape);
```
Maybe it would be slightly easier to follow the code if just using `push_back` on the `result_shape` instead?

amjames replied:
Agreed. I will change that.

wolfv commented on include/xtensor-blas/xlinalg.hpp:
```diff
@@ -1514,6 +1514,172 @@ namespace linalg
+        }
+
+        // Move the axes to sum over to the end of a
+        std::vector<std::size_t> newaxes_a;
```
same remark as above with `xt::dynamic_shape`

wolfv commented on include/xtensor-blas/xlinalg.hpp:
```diff
@@ -1514,6 +1514,172 @@ namespace linalg
+                newax_a_inserter = i;
+            }
+        }
+        for (auto a_ax_it = ax_a.begin(); a_ax_it != ax_a.end(); ++a_ax_it)
```
could also use range based for loop `for (const auto& el : ax_a) ...`

amjames replied:
Yes, the more verbose loop structure is holdover from an implementation that was not correct. I will change this.

wolfv commented on include/xtensor-blas/xlinalg.hpp:
```diff
@@ -1514,6 +1514,172 @@ namespace linalg
+                newax_b_inserter = i;
+            }
+        }
+        // not 100% sure why, but If I don't eval this I get the wrong result.
```
That sounds like a problem in ... `dot` maybe? I will have to look into it but for now we can keep the `eval` I guess.

wolfv commented:
We've internally discussed a couple of times how to do a "real" einsum. But we have not gotten around to actually do it. It would be super cool to have it in xtensor-blas! 

At some point we'd like to make one as part of core-xtensor, based on xsimd etc., but that is going to be way ... later.

wolfv commented:
This (limited) implementation of einsum looks fairly simple: https://github.com/jackkamm/einsum2 

But yeah, let's open a seperate issue for that, or discuss on gitter ( https://gitter.im/QuantStack/Lobby )

amjames commented on test/test_tensordot.cpp:
```diff
@@ -186,4 +186,68 @@ namespace xt
+
+    // }
+
+    TEST(xtensordot, reducing_dim_strided_view)
```
This test, and the commented out one above both cause the seg-fault.

amjames commented:
After looking this over with fresh eyes I noticed that none of my tests covered cases where views are provided as input. I added a few, and two of those tests I have added are causing a segfault. I am working on tracking this down.

amjames commented:
I was able to fix the tests on `strided_view`s. Which had the side effect of making the use of `eval` unnecessary.

wolfv commented on include/xtensor-blas/xlinalg.hpp:
```diff
@@ -1514,6 +1514,174 @@ namespace linalg
+            b_mat.reshape({sum_len, keep_b_len});
+            result = dot(a_mat, b_mat);
+            if(result_shape.empty()){
+              result.reshape({1});
```
Just some nitpicking regardign formatting:

- 4 spaces indennt, space after if and squiggly bracket on newline:

```
if (result_shape.empty())
{
    result.reshape({1});
}
else
{
...
}
```

wolfv commented on include/xtensor-blas/xlinalg.hpp:
```diff
@@ -1514,6 +1514,174 @@ namespace linalg
+        }
+        for (auto& a_ax_it : ax_a)
+        {
+          newaxes_a.push_back(a_ax_it);
```
formatting

wolfv commented on test/test_tensordot.cpp:
```diff
@@ -0,0 +1,254 @@
+{
+    TEST(xtensordot, outer_product)
+    {
+      xarray<double> a = xt::ones<double>({3,3,3});
```
indentation

tdegeus commented:
The docs of `xtensor` should probably be update too.

amjames commented:
@tdegeus would that just be adding `tensordot` to one of the tables at the end of the [numpy](https://github.com/QuantStack/xtensor/blob/master/docs/source/numpy.rst) section? 

Also the travis failure appears to be a connectivity hiccup that last commit just addressed the formatting issues raised by @wolfv, and only 1/13 builds failed while trying to `conda install openblas`

tdegeus commented:
I guess yes.

SylvainCorlay commented:
@amjames thank you for this contribution! This is really great work, especially for a first contribution.

Regarding the xtensor docs, this is not a blocker to get this one in since it is a different package.

wolfv commented:
@amjames last thing would be to add tensordot to the rst doc file in xtensor-blas https://github.com/QuantStack/xtensor-blas/blob/master/docs/source/reference.rst

Then I'll merge. 

Or if you don't have the time to do that right now, we can merge and then open an issue for this.

wolfv commented:
Looks great! I'll wait until the tests pass, and then merge!

wolfv commented:
congrats!
